### ISSUE TITLE
Rounding issue with fpToSBV

### ISSUE THREAD

**Issue by Btraube** [cvc5/cvc5#10997]
**Describe the bug**
The 'cvc5.pythonic. fpToSBV ( rm , x , s , ctx = None )'  function has a rounding issue. There seems to be a cutoff of all values after the decimal place when using this function. 

**cvc5 version/commit**: This is cvc5 version 1.1.3-dev.240.148acf9a0 [git 148acf9a0 on branch main] compiled with GCC version 11.4.0  
**Operating system**: Ubuntu 22.04.4 LTS 

```
fp1 = FPVal(5.5555)

simplify(fp1)
1.388875*(2**2)    # 5.5555

simplify(fpSqrt(RNE(), fp1))
1.1785054094063379*(2**1)   # 2.35701081881

simplify(fpToSBV(RNE(), fpSqrt(RNE(), fp1), BitVecSort(64)))
2

simplify(fpToSBV(RNA(), fpSqrt(RNE(), fp1), BitVecSort(64)))
2

simplify(fpToSBV(RTP(), fpSqrt(RNE(), fp1), BitVecSort(64)))
3

simplify(fpToSBV(RTN(), fpSqrt(RNE(), fp1), BitVecSort(64)))
2

simplify(fpToSBV(RTZ(), fpSqrt(RNE(), fp1), BitVecSort(64)))
2
```

**Comment by martin-cs** [comment]
( Cross reference #10962  )

**Comment by martin-cs** [comment]
Yes, it does.  That's the point.  There are three "round to integer" functions depending on whether you want a floating-point number, an unsigned bit-vector or signed bit-vector as the output.  I think this might be a documentation bug.  What output do you want it to give?

**Comment by Btraube** [comment]
Apologies if I have misunderstood something. I'm just coming over from using Z3 as our main library and am just looking for the corresponding CVC5 function. 

I'm looking for the output to be an IEEE BitVector representation of a floating-point value. Z3 has the function 'fpToIEEEBV(x)' where x is a floating-point constant. I thought that the fpToSBV was one of the two corresponding CVC5 functions to achieve that, but then ran into the issue of losing everything past the decimal point.

**Comment by aniemetz** [comment]
@Btraube it seems that we do not support the function `fpToIEEEBV(x)` (@alex-ozdemir is this an oversight or on purpose?). You can query sign, significand and exponent of a floating point value as documented here:
https://cvc5.github.io/docs-ci/docs-main/api/python/pythonic/fp.html#cvc5.pythonic.FPNumRef
Does this help?

**Comment by aniemetz** [comment]
It seems that this discussion is related: https://github.com/cvc5/cvc5/discussions/9277

**Comment by alex-ozdemir** [comment]
Hi folks,

Yes, we don't have `fpToIEEEBV` in the Pythonic interface (yet). In fact, we don't have the equivalent functionality in any of our interfaces. However, if you have a *constant* floating-point value, then `FPNumRef` methods that Aina refers to can give you the *constant* sign bit, exponent, etc.

To allow symbolic conversions from floating point values to their IEEE bit-vector representations, we'd have to start by making some of [these kinds](https://github.com/cvc5/cvc5/blob/main/src/theory/fp/kinds#L284-L300) public in cvc5's C++ API. Then, we could propagate this information out to the other APIs too.

**Comment by Btraube** [comment]
Thank you all for the clarification!

Apologies for making this an issue when there was a related discussion thread previously. I'll attempt to incorporate some of the potential work arounds from that other discussion post and I'll mess around with the `FPNumRef` method. I'll be keeping an eye out for the floating point to IEEE method in the python API! Out of curiosity, would you happen to know a timeline for that implementation? I absolutely understand if you don't have an idea for that yet. 

Again, thanks for your help with this!

**Comment by martin-cs** [comment]
> Hi folks,
> 
> Yes, we don't have `fpToIEEEBV` in the Pythonic interface (yet). In fact, we don't have the equivalent functionality in any of our interfaces. However, if you have a _constant_ floating-point value, then `FPNumRef` methods that Aina refers to can give you the _constant_ sign bit, exponent, etc.
> 
> To allow symbolic conversions from floating point values to their IEEE bit-vector representations, we'd have to start by making some of [these kinds](https://github.com/cvc5/cvc5/blob/main/src/theory/fp/kinds#L284-L300) public in cvc5's C++ API. Then, we could propagate this information out to the other APIs too.

I would suggest against making those kinds public, they are very very much internal glue to get SymFPU to work.  They will also not solve your problem ( as @aniemetz noted, in #9277 ) because `fptoIEEEBV` is not a total function.  Worse they may wind up with you getting the fixed bit pattern that is used in SymFPU and so making an internal implementation choice visible at theorem level.  Plus I suspect it will be incompatible with Z3.

@Btraube : I've been trying to create time to work on SymFPU all year, now we are approaching the end of teaching it will hopefully happen!

**Comment by martin-cs** [comment]
Hi all.  I think the right thing to do is to implement a two argument `fpToIEEEBV` function.  The first argument is the float to convert, the second is what to return in the case of a NaN.  This is inline with how other float partial functions are handled.  That way if you want z3-like, x86-like, ARM-like, RISC-like, etc. behaviour you can pick it and we avoid building that into the solver.

@Btraube would that solve your problem?

**Comment by Btraube** [comment]
Hello! I appreciate you getting back to me on this. I think that sounds like a great solution and that would indeed solve the problem I have run into.

**Comment by aniemetz** [comment]
Sorry for the late response, all. @martin-cs I don't see the problem with having a function that converts an FP value to the three string representations of its IEEE BV representation. We actually have this functionality in Bitwuzla's API (https://github.com/bitwuzla/bitwuzla/blob/main/include/bitwuzla/cpp/bitwuzla.h#L833). Users can always query if the FP value is NaN, and Z3 also only returns one single pattern for NaN: https://github.com/Z3Prover/z3/blob/master/src/api/python/z3/z3.py#L10934.

**Comment by aniemetz** [comment]
OK, so after double checking with what Z3 supports (it does support applying`fpToIEEEBV` to symbolic terms), we decided to not support this operator as we cannot reason about this symbolically.
